### PR TITLE
Visit `ImplItem` in `dead_code` lint

### PR DESCRIPTION
**Pull request by varkor**
Fixes https://github.com/rust-lang/rust/issues/47131.

### CODE REVIEW

File: src/librustc/middle/dead.rs [hunk 1]
```diff
@@ -172,6 +172,7 @@ impl<'a, 'tcx> MarkSymbolVisitor<'a, 'tcx> {
                     | hir::ItemKind::Ty(..)
                     | hir::ItemKind::Static(..)
                     | hir::ItemKind::Existential(..)
+                    | hir::ItemKind::Impl(..)
                     | hir::ItemKind::Const(..) => {
                         intravisit::walk_item(self, &item);
                     }
```
Comment (Centril) [inline src/librustc/middle/dead.rs]: Remove `_ => ()` and replace with "exhaustive" match to avoid future problems?

Comment (petrochenkov) [reply src/librustc/middle/dead.rs]: I'm not actually sure why this is not
```rust
_ => {
    intravisit::walk_item(self, &item);
}
```
If that doesn't break anything, then `walk_item` can be moved out of the `match`.

**Comment by varkor** [comment]
I've made the `dead_code` lint walk all items by default. I had to exclude `ItemKind::ForeignMod`, which caused a false negative and explicitly ignore a few `Def`s, which were now visited, but otherwise there weren't any problems. This should lower the chance an item is accidentally missed now.

**Pull request merged**